floodlight/core/events.py

```diff
@@ -61,3 +95,148 @@ def custom(self):
+        if not (definitions[col]["value_range"][0] <= check_col).all():
+            return False
+        if not (check_col <= definitions[col]["value_range"][1]).all():
+            return False
+
+        # all checks passed
+        return True
+
+    def add_frameclock(self, framerate: int):
+        """Add the column "frameclock", computed as the rounded multiplication of
+        gameclock and framerate, to the inner events DataFrame.
+
+        Parameters
+        ----------
+        framerate: int
+            Temporal resolution of data in frames per second/Hertz.
+        """
+        frameclock = np.full((len(self.events)), -1, dtype=int)
+        frameclock[:] = np.floor(self.events["gameclock"].values * framerate)
+        self.events["frameclock"] = frameclock
+
+    def find(
+        self, conditions: Tuple[str, Any] or List[Tuple[str, Any]]
+    ) -> pd.DataFrame:
+        """Returns a DataFrame containing all entries from the inner events DataFrame
+         that satisfy all given conditions.
+
+        Parameters
+        ----------
+        conditions: Tuple or List of Tuples
+            Conditions used to filter the columns. Can be either a single Tuple or a
+            List of Tuples. If given as a List, all conditions are enforced. Conditions
+            can be of the form (column, value) or of the form (column, (min, max)). In
+            case of the former, the specified column of the inner events DataFrame is
+            searched for entries equal to value. In case of the latter, the specified
+            column is searched for values in the range (min, max).
+            entries.
+
+        Returns
+        -------
+        filtered_events: pd.DataFrame
+            A copy of the inner events DataFrame where all entries fulfill all criteria
+            specified in conditions. The DataFrame can be empty if no entry with the
+            given value is found in the column.
+        """
+        filtered_events = self.events
+
+        # if single condition transform to list of conditions
+        if not isinstance(conditions, list):
+            conditions = [conditions]
+
+        for condition in conditions:
+            if condition[1] is None:
+                filtered_events = filtered_events[filtered_events[condition[0]].isna()]
+            else:
+                # check if a value or a value range is given
+                if isinstance(condition[1], list) or isinstance(condition[1], tuple):
+                    filtered_events = filtered_events[
+                        filtered_events[condition[0]] >= condition[1][0]
+                    ]
+                    filtered_events = filtered_events[
+                        filtered_events[condition[0]] <= condition[1][1]
+                    ]
+
+                else:
+                    filtered_events = filtered_events[
+                        filtered_events[condition[0]] == condition[1]
+                    ]
```

Comment:
```suggestion
        # loop through and filter by conditions
        for column, value in conditions:
        
            # only drop None's and NaN's from columns without a value specified
            if value is None:
                filtered_events = filtered_events[filtered_events[column].isna()]
                
            # check if a single value or a value range is given
            else:
            
                # value range: filter by minimum and maximum value
                if isinstance(value, (list, tuple)):
                    min_val, max_val = value
                    filtered_events = filtered_events[filtered_events[column] >= min_val]
                    filtered_events = filtered_events[filtered_events[column] <= max_val]
                    
                # single value: filter by that value
                else:
                    filtered_events = filtered_events[filtered_events[column] == value]
```

Reply:
This is is a good example how panda's syntax together with too few comments (for my taste) quickly becomes very hard to understand code. I've added a different approach that I hope is a little bit easier to read

Reply:
I agree that the comments improve readability, however, the comment at line 227 may be misleading. 

The idea of this line is to return the columns where the inner events DataFrame has the value None. 
E.g. self.find(('outcome', None)) should return all rows from the DataFrame where the outcome is None. I will replace it by:
# if the value is None filter for all entries with None, NaN or NA